[Utils] Identity map module-level debug info on first use in CloneFunction*

artempyanykh opened the pull request:
Stacked PRs:
 * #118630
 * #118629
 * #118628
 * __->__#118627


--- --- ---

### [Utils] Identity map module-level debug info on first use in CloneFunction*


Summary:
To avoid cloning module-level debug info (owned by the module rather
than the function), CloneFunction implementation used to eagerly
identity map such debug info into ValueMap's MD map. In larger modules
with meaningful volume of debug info this gets very expensive.

By passing such debug info metadata via an IdentityMD set for the
ValueMapper to map on first use, we get several benefits:

1. Mapping metadata is not cheap, particularly because of tracking. When
   cloning a Function we identity map lots of global module-level
   metadata to avoid cloning it, while only a fraction of it is actually
   used by the function. Mapping on first use is a lot faster for
   modules with meaningful amount of debug info.

2. Eagerly identity mapping metadata makes it harder to cache
   module-level data (e.g. a set of metadata nodes in a \a DICompileUnit).
   With this patch we can cache certain module-level metadata
   calculations to speed things up further.

Anecdata from compiling a sample cpp file with full debug info shows that this moderately speeds up
CoroSplitPass which is one of the heavier users of cloning:

|                 | Baseline | IdentityMD set |
|-----------------|----------|----------------|
| CoroSplitPass   | 306ms    | 221ms          |
| CoroCloner      | 101ms    | 72ms           |
| Speed up        | 1x       | 1.4x           |

Test Plan:
ninja check-llvm-unit
ninja check-llvm

felipepiovezan commented on llvm/lib/Transforms/Utils/CloneFunction.cpp:
```diff
@@ -322,16 +317,21 @@ void llvm::CloneFunctionInto(Function *NewFunc, const Function *OldFunc,
-  const auto RemapFlag = ModuleLevelChanges ? RF_None : RF_NoModuleLevelChanges;
+  // Current implementation always upgrades from local changes to module level
+  // changes due to the way metadata cloning is done. See
+  // BuildDebugInfoToIdentityMap for more details.
```
I can't find `BuildDebugInfoToIdentityMap` in this patch

felipepiovezan replied:
Also note how this comment reads: it talks about "upgrading local changes to module level changes", but there is nothing about the upgrade in the final code (as opposed to the  old code).  In other words, this comment is making a temporal reference to code that used to exist, which is not very helpful for readers.

We can phrase it as: "Cloning is always a Module level operation, since Metadata needs to be cloned".

artempyanykh replied:
> I can't find `BuildDebugInfoToIdentityMap` in this patch

Yes, sorry, this was renamed to `FindDebugInfoToIdentityMap`.

--

Updated the comment with the suggestion.

felipepiovezan commented on llvm/lib/Transforms/Utils/CloneFunction.cpp:
```diff
@@ -152,61 +152,54 @@ DISubprogram *llvm::CollectDebugInfoForCloning(const Function &F,
-    // Turn on module-level changes, since we need to clone (some of) the
-    // debug info metadata.
+    // Even if Changes are local only, we turn on module-level changes, since we
+    // need to clone (some of) the debug info metadata.
```
Similar to my other comment, is this comment needed? It is referring to code that used to exist, not to the code as it exists now

artempyanykh replied:
Good call! Agree this whole comment block doesn't add clarity. Removed it.

felipepiovezan reviewed (commented):
Left some initial comments, but I would like to test this commit on the swift compiler, since it makes heavy use of metadata cloning and could reveal potential issues. I'll do this asap

artempyanykh commented:
> ... I would like to test this commit on the swift compiler, since it makes heavy use of metadata cloning and could reveal potential issues. I'll do this asap

Thanks @felipepiovezan! Out of curiosity, is this about an open source version of swiftlang/swift and swiftlang/llvm-project?

felipepiovezan commented:
> > ... I would like to test this commit on the swift compiler, since it makes heavy use of metadata cloning and could reveal potential issues. I'll do this asap
> 
> Thanks @felipepiovezan! Out of curiosity, is this about an open source version of swiftlang/swift and swiftlang/llvm-project?

Yup! https://github.com/swiftlang/llvm-project/tree/swift/release/6.1

felipepiovezan commented:
I've tested this patch and it seems good, please ping me when you've addressed the other comments!

artempyanykh commented:
> I've tested this patch and it seems good

Wonderful! Thanks for checking @felipepiovezan

> Please ping me when you've addressed the other comments!

Will do! (most likely early next week)

artempyanykh commented:
@felipepiovezan pushed an update.

artempyanykh commented:
@felipepiovezan a gentle ping that this is ready for another look.

felipepiovezan commented:
> @felipepiovezan a gentle ping that this is ready for another look.

I'll have a look as soon as I get to the office tomorrow morning!

felipepiovezan commented on llvm/include/llvm/Transforms/Utils/Cloning.h:
```diff
@@ -242,13 +244,12 @@ DISubprogram *CollectDebugInfoForCloning(const Function &F,
-                         DISubprogram *SPClonedWithinModule);
+/// Based on \p Changes and \p DIFinder populate \p MD with debug info that
+/// needs to be identity mapped during Metadata cloning.
+void FindDebugInfoToIdentityMap(MetadataSetTy &MD,
```
Shouldn't MD be a return value instead of an out parameter? It's unclear why it has to be an outparameter from this description and usage of this function.

felipepiovezan commented on llvm/include/llvm/Transforms/Utils/ValueMapper.h:
```diff
@@ -136,6 +138,18 @@ inline RemapFlags operator|(RemapFlags LHS, RemapFlags RHS) {
+///    with meaningful amount of debug info.
+/// 3. Eagerly identity mapping metadata makes it harder to cache module-level
+///    data (e.g. a set of metadata nodes in a \a DICompileUnit).
+///
```
I would not place a comment about Cloning functions inside of the value mapper header; this comment is mostly relevant to "Cloning" users of this abstraction. We would better express this by moving this comment to `Cloning.h`. 

In ValueMapper.h, we should add a brief comment talking about when _users_ of this class should provide an IdentityMD and what it does. In other words, here we should provide guidance for people looking to use ValueMapper. Something short and descriptive like:

> If an IdentityMD set is optionally provided, all Metadata inside this set will be mapped onto itself.

artempyanykh replied:
This is fair! I moved the NOTE to `Cloning.h/FindDebugInfoToIdentityMap`. I don't think there's a single best place in Cloning.h to attach this NOTE to, but FindDebugInfoToIdentityMap feels like a reasonable one.

felipepiovezan commented on llvm/lib/Transforms/Utils/ValueMapper.cpp:
```diff
@@ -899,6 +901,14 @@ std::optional<Metadata *> Mapper::mapSimpleMetadata(const Metadata *MD) {
+  // larger than what will actually get used.
+  if (IdentityMD && IdentityMD->contains(MD)) {
+    return getVM().MD()[MD] = TrackingMDRef(const_cast<Metadata *>(MD));
+  }
```
[Don’t Use Braces on Simple Single-Statement Bodies of if/else/loop Statements](https://llvm.org/docs/CodingStandards.html#id60)

felipepiovezan reviewed (approved):
LGTM, but please address the smaller issues before merging!

artempyanykh commented:
Thank you @felipepiovezan, will do!

artempyanykh committed:
[Utils] Identity map module-level debug info on first use in CloneFunction*

Summary:
To avoid cloning module-level debug info (owned by the module rather
than the function), CloneFunction implementation used to eagerly
identity map such debug info into ValueMap's MD map. In larger modules
with meaningful volume of debug info this gets very expensive.

By passing such debug info metadata via an IdentityMD set for the
ValueMapper to map on first use, we get several benefits:

1. Mapping metadata is not cheap, particularly because of tracking. When
   cloning a Function we identity map lots of global module-level
   metadata to avoid cloning it, while only a fraction of it is actually
   used by the function. Mapping on first use is a lot faster for
   modules with meaningful amount of debug info.

2. Eagerly identity mapping metadata makes it harder to cache
   module-level data (e.g. a set of metadata nodes in a \a DICompileUnit).
   With this patch we can cache certain module-level metadata
   calculations to speed things up further.

Anecdata from compiling a sample cpp file with full debug info shows that this moderately speeds up
CoroSplitPass which is one of the heavier users of cloning:

|                 | Baseline | IdentityMD set |
|-----------------|----------|----------------|
| CoroSplitPass   | 306ms    | 221ms          |
| CoroCloner      | 101ms    | 72ms           |
| Speed up        | 1x       | 1.4x           |

Test Plan:
ninja check-llvm-unit
ninja check-llvm

stack-info: PR: https://github.com/llvm/llvm-project/pull/118627, branch: users/artempyanykh/fast-coro-upstream/8

artempyanykh commented:
Rebase
